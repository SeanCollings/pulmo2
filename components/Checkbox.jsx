import React from 'react';
import {
  StyleSheet,
  View,
  TouchableOpacity,
  Animated,
  Text,
} from 'react-native';
import { MaterialCommunityIcons } from '@expo/vector-icons';

import { useTheme } from '../hooks/useTheme';

const CHECKBOX_SIZE = 28;

const Checkbox = ({ state, onChange, label }) => {
  const theme = useTheme();
  const animatedScale = new Animated.Value(state ? 1 : 0.01);

  const checkboxHandler = () => {
    Animated.spring(animatedScale, {
      toValue: state ? 0.01 : 1,
      bounciness: 10,
      speed: 80,
    }).start(({ finished }) => {
      if (finished) onChange();
    });
  };

  const opacity = theme.DARK ? 0.87 : 1;
  const animatedStyleScale = {
    transform: [{ scale: animatedScale }],
  };

  return (
    <TouchableOpacity
      testID="checkboxButton"
      activeOpacity={1}
      onPress={checkboxHandler}
    >
      <View style={styles.container}>
        <View style={styles.checkContainer}>
          <MaterialCommunityIcons
            name={'checkbox-blank-outline'}
            color={theme.DARK ? theme.DARK_GRAY : theme.BORDER}
            size={CHECKBOX_SIZE}
            style={{ opacity }}
          />
          <Animated.View
            style={{
              ...styles.checkStyle,
              ...animatedStyleScale,
            }}
          >
            <MaterialCommunityIcons
              name={'check-bold'}
              color={theme.DARK ? theme.BORDER : theme.TEXT}
              size={CHECKBOX_SIZE - 12}
              style={{ opacity }}
            />
          </Animated.View>
        </View>
        <View style={{ ...styles.labelContainer, paddingLeft: label ? 10 : 0 }}>
          <Text style={{ ...styles.textStyle, color: theme.TEXT, opacity }}>
            {label}
          </Text>
        </View>
      </View>
    </TouchableOpacity>
  );
};

Checkbox.defaultProps = {
  state: false,
  onChange: () => null,
};

const styles = StyleSheet.create({
  container: { flexDirection: 'row' },
  checkContainer: { justifyContent: 'center', alignItems: 'center' },
  checkStyle: {
    position: 'absolute',
  },
  labelContainer: { justifyContent: 'center' },
  textStyle: {
    fontFamily: 'tit-light',
    fontSize: 16,
    paddingBottom: 2,
  },
});

export default Checkbox;
