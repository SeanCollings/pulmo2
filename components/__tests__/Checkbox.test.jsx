import React from 'react';
import { act, create } from 'react-test-renderer';
import { cleanup, fireEvent } from 'react-native-testing-library';
import Checkbox from '../Checkbox';

jest.useFakeTimers();
jest.mock('react-native/Libraries/Animated/src/NativeAnimatedHelper');

describe('Checkbox - unit test', () => {
  let component;
  const label = 'Checkbox';

  afterEach(() => {
    cleanup();
    jest.clearAllMocks();
  });

  test('should render in unchecked state', async () => {
    await act(async () => {
      component = create(<Checkbox state={false} label={label} />);
    });
    expect(component).toMatchSnapshot();
  });

  test('should render in checked state', async () => {
    await act(async () => {
      component = create(<Checkbox state label={label} />);
    });
    expect(component).toMatchSnapshot();
  });

  test('should handle checkbox toggle', async () => {
    await act(async () => {
      component = create(<Checkbox state label={label} />);
    });

    const checkboxButton = component.root.find(
      ({ props }) => props.testID === 'checkboxButton'
    );

    fireEvent.press(checkboxButton);
  });
});
