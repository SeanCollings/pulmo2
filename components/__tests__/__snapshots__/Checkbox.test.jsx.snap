// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Checkbox - unit test should render in checked state 1`] = `
<View
  accessible={true}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "opacity": 1,
    }
  }
  testID="checkboxButton"
>
  <View
    style={
      Object {
        "flexDirection": "row",
      }
    }
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "justifyContent": "center",
        }
      }
    >
      <Text
        allowFontScaling={false}
        style={
          Array [
            Object {
              "color": "#CCC",
              "fontSize": 28,
            },
            Object {
              "opacity": 1,
            },
            Object {
              "fontFamily": "material-community",
              "fontStyle": "normal",
              "fontWeight": "normal",
            },
            Object {},
          ]
        }
      >
        
      </Text>
      <View
        style={
          Object {
            "position": "absolute",
            "transform": Array [
              Object {
                "scale": 1,
              },
            ],
          }
        }
      >
        <Text
          allowFontScaling={false}
          style={
            Array [
              Object {
                "color": "#3e4b57",
                "fontSize": 16,
              },
              Object {
                "opacity": 1,
              },
              Object {
                "fontFamily": "material-community",
                "fontStyle": "normal",
                "fontWeight": "normal",
              },
              Object {},
            ]
          }
        >
          ﹮
        </Text>
      </View>
    </View>
    <View
      style={
        Object {
          "justifyContent": "center",
          "paddingLeft": 10,
        }
      }
    >
      <Text
        style={
          Object {
            "color": "#3e4b57",
            "fontFamily": "tit-light",
            "fontSize": 16,
            "opacity": 1,
            "paddingBottom": 2,
          }
        }
      >
        Checkbox
      </Text>
    </View>
  </View>
</View>
`;

exports[`Checkbox - unit test should render in unchecked state 1`] = `
<View
  accessible={true}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "opacity": 1,
    }
  }
  testID="checkboxButton"
>
  <View
    style={
      Object {
        "flexDirection": "row",
      }
    }
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "justifyContent": "center",
        }
      }
    >
      <Text
        allowFontScaling={false}
        style={
          Array [
            Object {
              "color": "#CCC",
              "fontSize": 28,
            },
            Object {
              "opacity": 1,
            },
            Object {
              "fontFamily": "material-community",
              "fontStyle": "normal",
              "fontWeight": "normal",
            },
            Object {},
          ]
        }
      >
        
      </Text>
      <View
        style={
          Object {
            "position": "absolute",
            "transform": Array [
              Object {
                "scale": 0.01,
              },
            ],
          }
        }
      >
        <Text
          allowFontScaling={false}
          style={
            Array [
              Object {
                "color": "#3e4b57",
                "fontSize": 16,
              },
              Object {
                "opacity": 1,
              },
              Object {
                "fontFamily": "material-community",
                "fontStyle": "normal",
                "fontWeight": "normal",
              },
              Object {},
            ]
          }
        >
          ﹮
        </Text>
      </View>
    </View>
    <View
      style={
        Object {
          "justifyContent": "center",
          "paddingLeft": 10,
        }
      }
    >
      <Text
        style={
          Object {
            "color": "#3e4b57",
            "fontFamily": "tit-light",
            "fontSize": 16,
            "opacity": 1,
            "paddingBottom": 2,
          }
        }
      >
        Checkbox
      </Text>
    </View>
  </View>
</View>
`;
